### PR TITLE
Add support for utf8 names on /v1/label/:name/values endpoint

### PR DESCRIPTION
**Pull request by ywwg**
Follow-on from https://github.com/prometheus/prometheus/pull/14943

Previously, the api was evaluating this regex to determine if the label
name was valid or not:

https://github.com/prometheus/common/blob/14bac55a992f7b83ab9d147a041e274606bdb607/model/labels.go#L94

However, I believe that the `IsValid()` function is what ought to be
used in the brave new utf8 era.

**Before**

```
$ curl localhost:9090/api/v1/label/host.name/values
{"status":"error","errorType":"bad_data","error":"invalid label name: \"host.name\""}
```

**After**

```
$ curl localhost:9090/api/v1/label/host.name/values
{"status":"success","data":["localhost"]}
```

It's very likely that I'm missing something here or you were already
planning to do this at some point but I just encountered this issue and
figured I'd give it a go.

### CODE REVIEW

**Comment by ywwg** [comment]
windows error seems to be spurious

**Review by bwplotka — DISMISSED** [review]
Nice, LGTM!

Should we do it in `release-3.0` branch first (if yes, rebase on that branch), so we do 3.0.1? cc @jan--f 

We need to put cap on how much we put into 3.0.x here. We will likely have multiple UTF-8 gaps.. I don't think it **must** go to 3.0.1, but I think it could (:

**Comment by ywwg** [comment]
@jan--f said to commit to main and then backport as necessary. I can set up that PR

**Review by beorn7 — CHANGES_REQUESTED** [review]

**Comment by beorn7** [comment]
FTR: I'm not sure if this should really be in v3.0.1. I would see it more as a missing feature than a real bug. (And I would like to get back to a regular release cadence ASAP, so that v3.1.0 is not far down the road.)
But _if_ we want to release this in v3.0.1, then let's first rebase this PR on top of the v3.0 release branch.

File: docs/querying/api.md
Comment (beorn7) [inline docs/querying/api.md:438]: Sorry for nitpicking again, but let's not use a "weird" label name in the first example. `job_name` is a weird because it's so common to have a `job` label. Why would somebody have `job_name`? If your intention is to have a label name with an underscore here, let' use one that is actually common like `http_status_code` (which could then be `http.status_code` in the example below to even allude to the different meaning of `.` and `_` in OTel).

File: docs/querying/api.md
Comment (beorn7) [inline docs/querying/api.md:452]: Strictly speaking, you can use non-legacy characters here as long as they are valid as URL components, can't we?

So I guess `http://localhost:9090/api/v1/label/job.name/values` would actually work?

If that's true, we only need the `U__` encoding for characters that don't work as URL components (in particular `/`).

File: docs/querying/api.md
Comment (beorn7) [inline docs/querying/api.md:453]: While it's handy to have a short description here, do we have a comprehensive explanation of this encoding method that we can link to?

**Review by beorn7 — CHANGES_REQUESTED** [review]
Sorry for more comments. Documenting is hard…

In any case, before merging this, let's decide if this should go into v3.0.1 or v3.1.0, see other comment. (If need be, we need to rebase this on top of the release branch.)

File: docs/querying/api.md
Comment (ywwg) [reply docs/querying/api.md:452]: I was thinking about that, but that might lead people to use HTML encoding: "job%20name" which is not supported. So while job.name would work, I'd prefer to make the reasoning simpler and just ask that people use encoding when the name is non-legacy-compatible.

File: docs/querying/api.md
Comment (ywwg) [reply docs/querying/api.md:453]: @beorn7 I can file an issue for the doc issue, is that enough to unblock this PR?

**Comment by jan--f** [comment]
Like beorn7 mentioned let's target `release-3.0` for a bug fix. Apologies if I added confusion and extra work with my previous statement to target `main`.

File: docs/querying/api.md
Comment (beorn7) [reply docs/querying/api.md:452]: By HTML encoding, you mean percent encoding, also known as URL encoding, don't you? I.e. https://en.wikipedia.org/wiki/Percent-encoding

I would expect that URL encoding is in fact supported, because that should just be dealt with by the Go HTTP libraries and the Prometheus code wouldn't even notice.

It might very well possible that we only need the `U__` encoding for a `/` in the label name. (At least that was the case long ago for label _values_ as they show up in URLs for pushing to the PGW. Of course, back then there was no `U__` encoding, so we elected for base64.)

Sorry for beating this point for so long, but it feels confusing to refer to the legacy Prometheus character set here when there is no technical reason for it.

File: docs/querying/api.md
Comment (beorn7) [reply docs/querying/api.md:453]: Linking to the Prometheus proposal is fine. I.e. link to https://github.com/prometheus/proposals/blob/main/proposals/2023-08-21-utf8.md#text-escaping (as we did for PGW, just checked that out). No issue required.

File: docs/querying/api.md
Comment (ywwg) [reply docs/querying/api.md]: I think I have rewritten it this way.

File: docs/querying/api.md
Comment (ywwg) [reply docs/querying/api.md]: done

**Comment by ywwg** [comment]
This should not block 3.0.1, but I believe it's ready to go in (pending getting the documentation right!)

**Review by beorn7 — APPROVED** [review]
It's actually good to have this in v3.0.1 because it is technically a breaking change (legacy label names could have the form `U__...`). I don't think it's relevant enough to call it out in the migration guide, though.

File: docs/querying/api.md
Comment (jan--f) [inline docs/querying/api.md:446]: I small nit that can be addressed in a follow up: `node` and `prometheus` are not `http_status_codes`.  The example is of course still valid but might give people pause. There is another instance of this further down.

**Review by jan--f — APPROVED** [review]
Could use a follow up, but shouldn't block this fix.

File: docs/querying/api.md
Comment (ywwg) [reply docs/querying/api.md:446]: ah good point

**Pull request merged**
